src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs

```diff
@@ -880,41 +881,62 @@ public static TimeZoneInfo FromSerializedString(string source)
+        public static ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones(bool skipSorting)
         {
             CachedData cachedData = s_cachedData;
 
             lock (cachedData)
             {
-                if (cachedData._readOnlySystemTimeZones == null)
+                if ((skipSorting ? cachedData._readOnlyUnsortedSystemTimeZones : cachedData._readOnlySystemTimeZones) is null)
                 {
-                    PopulateAllSystemTimeZones(cachedData);
-                    cachedData._allSystemTimeZonesRead = true;
+                    if (!cachedData._allSystemTimeZonesRead)
+                    {
+                        PopulateAllSystemTimeZones(cachedData);
+                        cachedData._allSystemTimeZonesRead = true;
+                    }
 
                     if (cachedData._systemTimeZones != null)
                     {
                         // return a collection of the cached system time zones
                         TimeZoneInfo[] array = new TimeZoneInfo[cachedData._systemTimeZones.Count];
                         cachedData._systemTimeZones.Values.CopyTo(array, 0);
 
-                        // sort and copy the TimeZoneInfo's into a ReadOnlyCollection for the user
-                        Array.Sort(array, static (x, y) =>
+                        if (!skipSorting)
                         {
-                            // sort by BaseUtcOffset first and by DisplayName second - this is similar to the Windows Date/Time control panel
-                            int comparison = x.BaseUtcOffset.CompareTo(y.BaseUtcOffset);
-                            return comparison == 0 ? string.CompareOrdinal(x.DisplayName, y.DisplayName) : comparison;
-                        });
-
-                        cachedData._readOnlySystemTimeZones = new ReadOnlyCollection<TimeZoneInfo>(array);
+                            // sort and copy the TimeZoneInfo's into a ReadOnlyCollection for the user
+                            Array.Sort(array, static (x, y) =>
+                            {
+                                // sort by BaseUtcOffset first and by DisplayName second - this is similar to the Windows Date/Time control panel
+                                int comparison = x.BaseUtcOffset.CompareTo(y.BaseUtcOffset);
+                                return comparison == 0 ? string.CompareOrdinal(x.DisplayName, y.DisplayName) : comparison;
+                            });
+
+                            // Always reset _readOnlyUnsortedSystemTimeZones even if it was initialized before. This prevents the need to maintain two separate cache lists in memory
+                            // and guarantees that if _readOnlySystemTimeZones is initialized, _readOnlyUnsortedSystemTimeZones is also initialized.
+                            cachedData._readOnlySystemTimeZones = cachedData._readOnlyUnsortedSystemTimeZones = new ReadOnlyCollection<TimeZoneInfo>(array);
```

Comment:
Doesn't this do an assignment by value of `new ReadOnlyCollection<TimeZoneInfo>(array);` into both `cachedData._readOnlySystemTimeZones` and `cachedData._readOnlyUnsortedSystemTimeZones`? Or is the ref of `new ReadOnlyCollection<TimeZoneInfo>(array)` being passed to both?
Am I right in thinking we are still using twice as much space as we really need to here?

Reply:
The line 

```C#
cachedData._readOnlySystemTimeZones = cachedData._readOnlyUnsortedSystemTimeZones = new ReadOnlyCollection<TimeZoneInfo>(array);
```

is equivalent to 

```C#
cachedData._readOnlyUnsortedSystemTimeZones = new ReadOnlyCollection<TimeZoneInfo>(array);
cachedData._readOnlySystemTimeZones = cachedData._readOnlyUnsortedSystemTimeZones 
```

So, it only creates one collection object and assigns it to both fields. Look at the following example to show you how the compiler handle it.

https://sharplab.io/#v2:EYLgHgbALAPgAgJgIwFgBQcAMACOSoDc66cAzLgtgMLboDe62Tu5cU2AsgBQCUtazbAwGDmAGQCWAZwAuAHjyYAfNgA20mUiIjR2SbIVJlajQm26mjC+tlJsAXhOzKjgHYBTAO56Nh5b3MLK10bTQA6AEEAEyiuACIJGXcAWyQ4nkCQ00iY+MSUhHTM7ABfdBKgA